lib/galaxy/authnz/custos_authnz.py

```diff
@@ -67,6 +67,44 @@ def __init__(self, provider, oidc_config, oidc_backend_config, idphint=None):
     def _decode_token_no_signature(self, token):
         return jwt.decode(token, audience=self.config["client_id"], options={"verify_signature": False})
 
+    def refresh(self, trans):
+        custos_authnz_token = self._get_custos_authnz_token(
+            trans.sa_session, trans.user.custos_auth[0].external_user_id, self.config["provider"]
+        )
+        if custos_authnz_token is None:
+            raise exceptions.AuthenticationFailed("cannot find authorized user while refreshing token")
+        if (custos_authnz_token.expiration_time - datetime.now()).total_seconds() > 60:
```

Comment:
The refresh condition for this appears to be different for the PSA case: https://github.com/galaxyproject/galaxy/pull/15300/files#diff-52d2ab08a1b269dab28ef1bb2feb00aef8eac9a200797994f34d1868b7de51c1R171

Should they be the same?

Reply:
The reason is that `CustosAuthnzToken` and `UserAuthnzToken` have different fields, in particular `CustosAuthnzToken` does not contain info about auth time, so I could not refresh custos tokens at their half-life (at least not without changing the database scheme).